dnf5-plugins/system_upgrade_plugin/system_upgrade.cpp

```diff
@@ -0,0 +1,517 @@
+    poweroff_after_arg->set_long_name("poweroff");
+    poweroff_after_arg->set_description("Power off the system after the operation is complete");
+    poweroff_after_arg->link_value(poweroff_after);
+
+    cmd.register_named_arg(poweroff_after_arg);
+}
+
+void SystemUpgradeRebootCommand::run() {
+    auto state = get_state();
+    check_state(state);
+    if (state.get_data().status != STATUS_DOWNLOAD_COMPLETE) {
+        throw libdnf5::cli::CommandExitError(1, M_("system is not ready for upgrade"));
+    }
+    if (std::filesystem::is_symlink(get_magic_symlink())) {
+        throw libdnf5::cli::CommandExitError(1, M_("upgrade is already scheduled"));
+    }
+
+    if (!std::filesystem::is_directory(get_datadir())) {
+        throw libdnf5::cli::CommandExitError(1, M_("data directory {} does not exist"), get_datadir().string());
+    }
+
+    std::filesystem::create_symlink(get_datadir(), get_magic_symlink());
+
+    state.get_data().status = STATUS_READY;
+    state.get_data().poweroff_after = poweroff_after->get_value();
+    state.write();
+
+    reboot(get_context(), poweroff_after->get_value());
+}
+
+void SystemUpgradeUpgradeCommand::set_argument_parser() {}
+
+void SystemUpgradeUpgradeCommand::configure() {
+    SystemUpgradeSubcommand::configure();
+    auto & ctx = get_context();
+
+    ctx.set_load_system_repo(true);
+    ctx.set_load_available_repos(Context::LoadAvailableRepos::ENABLED);
+
+    auto state = get_state();
+    check_state(state);
+
+    // Set same set of enabled/disabled repos used during `system-upgrade download`
+    for (const auto & repo_id : state.get_data().enabled_repos) {
+        ctx.setopts.emplace_back(repo_id + ".enabled", "1");
+    }
+    for (const auto & repo_id : state.get_data().disabled_repos) {
+        ctx.setopts.emplace_back(repo_id + ".disabled", "1");
+    }
+
+    // Don't try to refresh metadata, we are offline
+    ctx.base.get_config().get_cacheonly_option().set(libdnf5::Option::Priority::PLUGINDEFAULT, "all");
+    // Don't ask any questions
+    ctx.base.get_config().get_assumeyes_option().set(libdnf5::Option::Priority::PLUGINDEFAULT, true);
+    // Override `assumeno` too since it takes priority over `assumeyes`
+    ctx.base.get_config().get_assumeno_option().set(libdnf5::Option::Priority::PLUGINDEFAULT, false);
+    // Upgrade operation already removes all element that must be removed.
+    // Additional removal could trigger unwanted changes in transaction.
+    ctx.base.get_config().get_clean_requirements_on_remove_option().set(
+        libdnf5::Option::Priority::PLUGINDEFAULT, false);
+    ctx.base.get_config().get_install_weak_deps_option().set(libdnf5::Option::Priority::PLUGINDEFAULT, false);
+}
+
+void SystemUpgradeUpgradeCommand::run() {
+    auto & ctx = get_context();
+
+    if (!std::filesystem::is_symlink(get_magic_symlink())) {
+        throw libdnf5::cli::CommandExitError(0, M_("Trigger file does not exist. Exiting."));
```

Comment:
Related issue: https://github.com/rpm-software-management/dnf-plugins-core/issues/512.

Reply:
Resolved in #1280: https://github.com/rpm-software-management/dnf5/pull/1280/commits/bd6b65bae7feb3d7c8d69285b4d7523b9ca30ca3